base/bundle.cpp

```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+#include "base/bundle.hpp"
+
+#include <functional>
+#include <list>
+
+#include "base/map_util.hpp"
+#include "base/status.hpp"
+
+namespace principia {
+namespace base {
+
+thread_local std::function<bool()> AbortRequested = [] { return false; };
+
+Bundle::TaskHandle::TaskHandle(std::shared_ptr<Task> task)
+    : task_(std::move(task)) {}
+
+Bundle::Bundle(int threads)
+    : terminate_on_empty_(false),
+      max_workers_(threads) {
+  workers_.reserve(max_workers_);
+}
+
+Status Bundle::Join() {
+  {
+    std::unique_lock<std::mutex> lock(lock_);
+    CHECK(!terminate_on_empty_);
+    terminate_on_empty_ = true;
+  }
+  tasks_not_empty_or_terminate_.notify_all();
+  for (auto& thread : workers_) {
+    thread.join();
+  }
+  std::unique_lock<std::mutex> lock(lock_);
```

Comment:
Not needed since no other thread is running at this point (hopefully you cannot call Add and succeed once you've started joining).